### DISCUSSION TITLE
Creating a single-file bundle

### DISCUSSION THREAD

**Discussion by dkarlovi** [symfony/symfony#43080] [Q&A] [answered]
I'm trying to create a single-file bundle which would autowire / autoconfigure a namespace.

Currently I have this:
```php
<?php

declare(strict_types=1);

namespace Import;

use Symfony\Component\Config\FileLocator;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Extension\Extension;
use Symfony\Component\DependencyInjection\Extension\ExtensionInterface;
use Symfony\Component\DependencyInjection\Loader;
use Symfony\Component\HttpKernel\Bundle\Bundle;

class ImportBundle extends Bundle
{
    public function createContainerExtension(): ExtensionInterface
    {
        return new class extends Extension
        {
            public function load(array $configs, ContainerBuilder $container): void
            {
                $loader = new Loader\PhpFileLoader($container, new FileLocator(__DIR__.'/../config'));
                $loader->import('config.php');
            }
            
            public function getAlias(): string
            {
                return 'import';
            }
        };
    }
}
```

and

```php
<?php

use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $container): void {
    $container = $container
        ->services()
            ->defaults()
                ->private()
                ->autoconfigure()
                ->autowire();

    $container
        ->load('Import\\', '../src/*')
        ->exclude('../src/{Model}');
};
```

Is it possible to merge these two files into one? I was looking at using the Callback loader, but it doesn't seem to work as I expect.

TIA.

**Comment by ro0NL** [comment]
Looks like ClosureLoader can't take a configurator. Perhaps a nice feature :) ref https://github.com/symfony/symfony/issues/42647 also

**Comment by stof** [comment]
If you want to keep things in a single file, your extension class will need to define its services using the lower-level API of the DI component (i.e. using the `ContainerBuilder` and its `Definition` instances) instead of using loaders that load config files (which by definition require using a separate file)

**Reply by dkarlovi** [reply]
OK, have an example?

**Reply by dkarlovi** [reply]
> loaders that load config files (which by definition require using a separate file)

There's `\Symfony\Component\DependencyInjection\Loader\ClosureLoader` which doesn't.

**Comment by yceruto** [comment, marked as answer]
This proposal https://github.com/symfony/symfony/pull/43701 could help here.

**Reply by dkarlovi** [reply]
Nice! How would it work?

**Reply by yceruto** [reply]
Take a look at the PR description, everything is there :)

**Comment by yceruto** [comment]
I just wanted to come back to this ticket to let you know that you can now use https://github.com/yceruto/micro-symfony for this goal (it is available for 5.4 so far).
